### ISSUE TITLE
Office API 1.1 getTaskByIndexAsync gets lost if more than 10(ish) Project Tasks

### ISSUE THREAD

**Issue by indieonly** [OfficeDev/office-js-docs#584]
Hello,
in the Code below, my Function buildTaskHierarchy below is never called where there are more than 10-11 Tasks in the Project; otherwise it works fine. This concerns Project 2016, which is needed for Office API 1.1.

I think it is clear that the Office API Async-Calls manager is being overloaded (this is almost visible in VisualStudio debugging) and gets lost after more than just 10 calls to getTaskByIndexAsync (multiplied by the number of Field values requested via getTaskFieldAsync -- in parallel -- in a loop).

Given that "in the current version of the JavaScript API for Office, the promises pattern only works with code for bindings in Excel spreadsheets and Word documents", it would appear that the Office API 1.1, despite providing the new Methods getTaskByIndexAsync and getTaskFieldAsync, does not actually permit the Developer to recover all necessary Task information (even if we are only talking about 3-4 Fields) for a real Project (i.e. one that has far more than 10 tasks), for which an Add-In might actually be of some value (if someone has a Project with 10 tasks, I don't imagine they need an Add-In).

I have just spent 3 days developing my Add-In -- which does not currently permit me to get all needed Project Task information en masse -- so can someone please tell me if this apparent issue with the API 1.1 is known, and will be fixed?

Cameron

...
    maxTaskIndex = asyncResult.value;
        for (var taskIndex = 1; taskIndex <= maxTaskIndex; taskIndex++) {
            // For every Task (Index), get its GUID and then get needed Fields for Task
            _projDoc.getTaskByIndexAsync(taskIndex, getTaskFieldsByGUID);
        }
...
function getTaskFieldsByGUID(asyncResult) {
    var taskGUID = asyncResult.value;
    var taskJSON = {};

        taskJSON = { "taskGUID": taskGUID };
        tasksArray.push(taskJSON);

        // Get Task Name -- then store in Task JSON
        _projDoc.getTaskFieldAsync(taskGUID,
                                    Office.ProjectTaskFields.Name,                      
                                    { asyncContext: { "taskGUID": taskGUID,
                                                      "taskFieldID": Office.ProjectTaskFields.Name } },
                                    pushTaskFieldIntoTaskJSON);
        // Get Task WBS (ID) -- then store in Task JSON
        _projDoc.getTaskFieldAsync(taskGUID,
                                    Office.ProjectTaskFields.WBS,                       
                                    { asyncContext: { "taskGUID": taskGUID,
                                                      "taskFieldID": Office.ProjectTaskFields.WBS } },
                                    pushTaskFieldIntoTaskJSON);
        // Get Task Work -- then store in Task JSON
        _projDoc.getTaskFieldAsync(taskGUID,
                                    Office.ProjectTaskFields.Work,
                                    { asyncContext: { "taskGUID": taskGUID,
                                                      "taskFieldID": Office.ProjectTaskFields.Work } },
                                    pushTaskFieldIntoTaskJSON);
... 
function pushTaskFieldIntoTaskJSON(asyncResult) {

        // Capture the current Task Name and Value in the Task JSON on success
        for (taskArrayIndex = 0; taskArrayIndex < tasksArray.length; taskArrayIndex++) {

            // Start by finding current Task in global Task Array (by GUID)
            if (tasksArray[taskArrayIndex].taskGUID == asyncResult.asyncContext.taskGUID) {
                // Current Task JSON Object has been found in global Task Array
                switch (asyncResult.asyncContext.taskFieldID) {
                    // Task Name
                    case Office.ProjectTaskFields.Name:
                        tasksArray[taskArrayIndex].taskName = asyncResult.value.fieldValue;
                        break;
                    // Task WBS (ID)
                    case Office.ProjectTaskFields.WBS:
                        tasksArray[taskArrayIndex].taskWBS = asyncResult.value.fieldValue;
                        break;
                    // Task Work (in hours)
                    case Office.ProjectTaskFields.Work:
                        tasksArray[taskArrayIndex].taskWork = asyncResult.value.fieldValue;
                        break;
                }
                break;         // leave FOR immediately
            }
        }   // FOR

        if (taskArrayIndex == (maxTaskIndex - 1)) {          
            // Is this the last Task? Array Index starts at 0 unlike maxTaskIndex
            if (typeof tasksArray[taskArrayIndex].taskName !== 'undefined'
                && typeof tasksArray[taskArrayIndex].taskWBS !== 'undefined'
                && typeof tasksArray[taskArrayIndex].taskWork !== 'undefined') {

                // Once the last Task Field Value has been returned for the last Task ...
                buildTaskHierarchy();
...

**Comment by indieonly** [comment]
The following Exception appears to be thrown consistently after the first 51 "Agave.HostCall.IssueCall" lines in the VisualStudio output box (which equates to supporting a maximum of 10 Tasks if you first get the GUID by Index, then get 4 fields per GUID as in my tests):

Exception was thrown at line 2953, column 13 in https://localhost:44378/Scripts/Office/1/project-win32-16.00.debug.js
0x80043200 - JavaScript runtime error: unknown exception

**Comment by keyur32** [comment]
Following up with our Project team on this issue. thanks

**Comment by kriede** [comment]
We are having the same issue with getTaskByIndexAsync: After 50 Tasks we get 5001 Internal Error on subsequent calls. Any news on this issue?

**Comment by indieonly** [comment]
Hi Kriede,
am not sure you will be getting a response any time soon.

You can Code around this obvious limitation by ensuring no more than 40-45 Async calls are made in parallel using JQuery $.Deferred (see tested Code example below). Having said that, my testing revealed a SHARP drop in performance as I got towards 40-45 parallel Async calls (tested by MS?), with around 5 parallel Async calls strangely performing much faster (in which case $.Deferred may not even be strictly necessary, but is more robust). So, keep your packets small for readability and strangely, better performance.

Best,
Cameron

function getTaskOtherFieldsThenNext(taskIndex, maxTaskIndex, deferAllTasks) {

    // Create a Defer Object for the current Task to be read and stored against a JSON Object
    var deferTask1 = $.Deferred();

    // Is current Task a Leaf (or the Project Task) ?    
    if (tasksArray[taskIndex].taskIsLeaf == true || taskIndex == 0) {      
        getTaskJSONOtherFields(tasksArray[taskIndex], deferTask1);

    } else {
        // Only need ******* for Leaves and Parent Task, so Resolve promise immediately for other tasks to continue
        deferTask1.resolve();
    }
            
    // Processing to be performed once Task Promise has been resolved
    $.when( deferTask1 ).done( function () {
        // getTaskJSONKeyFields has finished its work for current Task. Go to next Tasks or finish
        if (taskIndex >= maxTaskIndex) {
            // If we have treated the last task, then 'All Tasks' Promise has been fulfilled
            return deferAllTasks.resolve();

        } else {
            // Kick off read for next Task (n + 1)
            taskIndex++;
            getTaskOtherFieldsThenNext(taskIndex, maxTaskIndex, deferAllTasks);
        }
    });
}
